### PR TITLE
Port MerkleChangeTracker to use arbitrary backend

### PR DESCRIPTION
**Pull request by CjS77**
This PR ports SWvHeerden's ChangeTracker to the new MMR crate to enable it to
be used with arbitrary backends and backing MMRs.  For example, you can build a
rewindable MMR with a pruned MMR based in LMDB with a memory-babcked
checkpoined MMR, or any combination of the two.

This PR adds the `MerkleChangeChecker` struct and supporting functions.  The
`ArrayLike` like was extended to include two new functions, `trunacate` and an
iterator-like function `for_each`.  For implementations that don't want to, or
can't support these new features, a `NotSupported` error can be returned.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [x] My change requires a change to the documentation.
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/mmr/src/change_tracker.rs [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+pub struct MerkleChangeTracker<D, BaseBackend, CpBackend>
+where
+    D: Digest,
+    BaseBackend: ArrayLike<Value = Hash>,
+{
+    base: MutableMmr<D, BaseBackend>,
+    mmr: PrunedMutableMmr<D>,
+    checkpoints: CpBackend,
+    // The hashes added since the last commit
+    current_hashes: Vec<Hash>,
```
Comment (mikethetike) [inline base_layer/mmr/src/change_tracker.rs]:
```suggestion
    current_additions: Vec<Hash>,
```
Not entirely sure if these are only additions, but it seems to be

File: base_layer/mmr/src/backend.rs [hunk 2]
```diff
@@ -39,6 +39,16 @@ pub trait ArrayLike {
...
     fn get_or_panic(&self, index: usize) -> &Self::Value;
+
+    /// Shortens the array, keeping the first len elements and dropping the rest. If this feature is not supported,
+    /// the function should return `Err(MerkleMountainRangeError:NotSupported)`
+    fn truncate(&mut self, _len: usize) -> Result<(), MerkleMountainRangeError> {
+        Err(MerkleMountainRangeError::NotSupported)
+    }
+
+    /// Execute the given closure for each value in the array
+    fn for_each<F>(&self, f: F) -> Result<(), MerkleMountainRangeError>
```
Comment (mikethetike) [inline base_layer/mmr/src/backend.rs]: This seems very specific to a call in the ChangeTracker. Perhaps it should live in its own interface, or perhaps only be implemented for `ArrayLike` of `Result<_>`?

File: base_layer/mmr/src/change_tracker.rs (hunk 1 above)
Comment (CjS77) [reply base_layer/mmr/src/change_tracker.rs]: Changed

File: base_layer/mmr/src/backend.rs (hunk 2 above)
Comment (CjS77) [reply base_layer/mmr/src/backend.rs]: I like the different trait idea. Negates need for stubbed "NotSupported" methods. Implemented.

**Pull request merged**